### PR TITLE
Added UnigueList

### PR DESCRIPTION
**Pull request by Christian-B**
This is a simple override of list that (I hope) intercepts any of the methods that could add elements to the list and silently removes any duplicates.

Differences to OrderedSet are.
1. Has the list interface including indexing
2. May well be slower on any method which changes it or check membership.
3. Probably smaller memory size

### CODE REVIEW

**Comment by Christian-B** [comment]
Much slower expesially on bigger lists!
UnigueList took 7.763265371322632 compared to OrderedSet of  0.024962663650512695

import random
import time
from spinn_utilities.unigue_list import UnigueList
from spinn_utilities.ordered_set import OrderedSet

data = [random.randint(1, 10000) for i in range(100000)]

start = time.time()
ul = UnigueList()
for foo in data:
    ul.append(foo)
print(len(ul))
print (time.time()-start)
start = time.time()
os = OrderedSet()
for foo in data:
    os.add(foo)
print(len(ul))
print (time.time()-start)

File: spinn_utilities/unigue_list.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+class UnigueList(list):
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:17]: Please change the name to `UniqueList`.
Filename must also change to `unique_list.py`.

File: spinn_utilities/unigue_list.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        :param p_object: something to add to the lsit
+        """
+        if p_object in self:
+            return
+        super(UnigueList, self).append(p_object)
+
+    def copy(self):
+        """
+        Make a copy/ clone of this list
+        :return: a new list
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:46]: You _need_ a blank line before this line.

File: spinn_utilities/unigue_list.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        :param iterable: iterable
+        """
+        for p_object in iterable:
+            self.append(p_object)
+
+    def insert(self, index, p_object):
+        """
+        insert object before index
+        
+        note:  If the object was already in the list it is removed and then
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:66]: Better to write as:
```
        .. note::
            If the object was already in the list it is removed and then
```
as that renders differently.

File: spinn_utilities/unigue_list.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        location found
+        :rtype: UnigueList
+        """
+        new_list = UnigueList(self)
+        new_list.extend(args[0])
+        return new_list
+
+    def __iadd__(self, *args, **kwargs):
+        """ inplace add which ignores duplicates
+        :return: a new list with a single copy in of all items in the first
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:93]: Blank line needed before this line

File: spinn_utilities/unigue_list.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        :param int index: (new) location for the object
+        :param p_object: object to add in
+        """
+        self.insert(args[0], args[1])
+
+    def __imul__(self, *args, **kwargs):
+        """"
+        Multiplication typicaly just adds duplicate copies so is ingnored
+        :rtype: UnigueList
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:116]: Blank line needed before this line

File: spinn_utilities/unigue_list.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        self.extend(args[0])
+        return self
+
+    def __setitem__(self, *args, **kwargs):
+        """
+        insert object before index
+
+        note:  If the object was already in the list it is removed and then
```
Comment (dkfellows) [inline spinn_utilities/unigue_list.py:104]: See previous comment about `.. note::`

File: spinn_utilities/set_list.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    """
+    A object that acts both as a Set and a List where when the rules different
+    set take preference.
+    """
+
+    __slots__ = ["as_set"]
+
+    def __init__(self, seq=()):
+        """
+        Creates a list like object and fills in the intial values igornoring
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:33]: Don't normally need a descriptive paragraph on an `__init__` method.

File: spinn_utilities/set_list.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        """
+        if p_object in self.as_set:
+            return
+        super(SetList, self).append(p_object)
+        self.as_set.add(p_object)
+
+    def copy(self):
+        """
+        Make a copy/ clone of this list
+        :return: a new list
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:57]: Blank line needed before this line.

File: spinn_utilities/set_list.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        :param iterable: iterable
+        """
+        for p_object in iterable:
+            self.append(p_object)
+
+    def insert(self, index, p_object):
+        """
+        insert object before index
+        
+        note:  If the object was already in the list it is removed and then
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:77]: See other comment about `.. note::`

File: spinn_utilities/set_list.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        location found
+        :rtype: UnigueList
+        """
+        new_list = SetList(self)
+        new_list.extend(args[0])
+        return new_list
+
+    def __iadd__(self, *args, **kwargs):
+        """ inplace add which ignores duplicates
+        :return: a new list with a single copy in of all items in the first
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:106]: Blank line needed before this line.

File: spinn_utilities/set_list.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+        :param int index: (new) location for the object
+        :param p_object: object to add in
+        """
+        self.insert(args[0], args[1])
+
+    def __imul__(self, *args, **kwargs):
+        """"
+        Multiplication typicaly just adds duplicate copies so is ingnored
+        :rtype: UnigueList
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:129]: Blank line needed before this line.

File: spinn_utilities/set_list.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        return self.as_set.issubset(*args, **kwargs)
+
+    def issuperset(self, *args, **kwargs):
+        """ Report whether this set contains another set. """
+        return self.as_set.issuperset(*args, **kwargs)
+
+    def pop(self, *args, **kwargs):
+        """
+        Remove and return an arbitrary set element.
+        Raises KeyError if the set is empty.
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:199]: Should be:
```
    :raises KeyError: if the set is empty.
```
and should have a blank line before it.

File: spinn_utilities/set_list.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+            raise KeyError('set is empty')
+        p_object = super(SetList, self).pop(*args, **kwargs)
+        self.as_set.remove(p_object)
+        return p_object
+
+    def remove(self, *args, **kwargs):
+        """
+        Remove an element from a set; it must be a member.
+
+        If the element is not a member, raise a KeyError.
```
Comment (dkfellows) [inline spinn_utilities/set_list.py:211]: Should be:
```
    :raises KeyError: If the element is not a member.
```

**Review by dkfellows — CHANGES_REQUESTED** [review]
Needs a bit of work.

**Comment by Christian-B** [comment]
While I agree with all of @dkfellows comments I will not fix as this implementations shows it is significantly slower than OrderedSet and also much slower than our SetList.
The larger the resulting set will get the worse the difference.

See test.py.

My suggestion is to close this PR.

**Comment by Christian-B** [comment]
too slow and not needed.

**Pull request closed without merging**